cli/internal/lockfile/npm_lockfile.go

```diff
@@ -0,0 +1,232 @@
+	nestedPath := fmt.Sprintf("%s/node_modules/%s", workspacePath, name)
+	if entry, ok := l.Packages[nestedPath]; ok {
+		return Package{
+			Key:     nestedPath,
+			Version: entry.Version,
+			Found:   true,
+		}, nil
+	}
+
+	// Next we check for a top level version of the package
+	hoistedPath := fmt.Sprintf("node_modules/%s", name)
+	if entry, ok := l.Packages[hoistedPath]; ok {
+		return Package{
+			Key:     hoistedPath,
+			Version: entry.Version,
+			Found:   true,
+		}, nil
+	}
+
+	return Package{}, nil
+}
+
+// AllDependencies Given a lockfile key return all (dev/optional/peer) dependencies of that package
+func (l *NpmLockfile) AllDependencies(key string) (map[string]string, bool) {
+	entry, ok := l.Packages[key]
+	if !ok {
+		return nil, false
+	}
+	deps := make(map[string]string, len(entry.Dependencies)+len(entry.DevDependencies)+len(entry.PeerDependencies)+len(entry.OptionalDependencies))
+	addDep := func(d map[string]string) {
+		for name := range d {
+			for _, possibleKey := range possibleNpmDeps(key, name) {
+				if entry, ok := l.Packages[possibleKey]; ok {
+					deps[possibleKey] = entry.Version
+					break
+				}
+			}
+		}
+	}
+
+	addDep(entry.Dependencies)
+	addDep(entry.DevDependencies)
+	addDep(entry.OptionalDependencies)
+	addDep(entry.PeerDependencies)
+
+	return deps, true
+}
+
+// Subgraph Given a list of lockfile keys returns a Lockfile based off the original one that only contains the packages given
+func (l *NpmLockfile) Subgraph(workspacePackages []turbopath.AnchoredSystemPath, packages []string) (Lockfile, error) {
+	prunedPackages := make(map[string]NpmPackage, len(packages))
+	for _, pkg := range packages {
+		if entry, ok := l.Packages[pkg]; ok {
+			prunedPackages[pkg] = entry
+		} else {
+			return nil, fmt.Errorf("No lockfile entry found for %s", pkg)
+		}
+	}
+	if rootEntry, ok := l.Packages[""]; ok {
+		prunedPackages[""] = rootEntry
+	}
+	for _, workspacePackages := range workspacePackages {
+		workspacePkg := workspacePackages.ToUnixPath().ToString()
+		if workspaceEntry, ok := l.Packages[workspacePkg]; ok {
+			prunedPackages[workspacePkg] = workspaceEntry
+		} else {
+			return nil, fmt.Errorf("No lockfile entry found for %s", workspacePkg)
+		}
+		// Each workspace package has a fake version of the package that links back to the original
+		// but is needed for dependency resolution.
+		for key, entry := range l.Packages {
+			if entry.Resolved == workspacePkg {
+				prunedPackages[key] = entry
+				break
+			}
+		}
+	}
+
+	return &NpmLockfile{
+		Name:    l.Name,
+		Version: l.Version,
+		// Forcible upgrade to version 3 since we aren't including backward compatibility capabilities
+		LockfileVersion: 3,
+		Requires:        l.Requires,
+		Packages:        prunedPackages,
+		// We omit dependencies since they are only used for supporting npm <=6
+	}, nil
+}
+
+// Encode encode the lockfile representation and write it to the given writer
+func (l *NpmLockfile) Encode(w io.Writer) error {
```

Comment:
curious why you chose `NpmLockfile` to be the receiver, rather than a standalone utility function? Seems like it would be the same implementation elsewhere too?

Reply:
`Encode` is part of the `Lockfile` interface and due to the different file formats there isn't really any shared behavior between the implementations.